### PR TITLE
Python Wrapper: Add examples to documentation + prettify

### LINKED ISSUE #7040: Python Wrapper: Add examples into the docstrings
Adding examples into our classes and methods docstrings will allow us to incorporate them into the documentation and provide better experience for users going over the docs

### LINKED ISSUE #7068: [BUG]: Python Wrapper: Documentation sidebar overflow
https://pydocs-lakefs.lakefs.io/lakefs.config.html#lakefs.config.ClientConfig

image

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7068
Closes #7040

## Change Description

- Improve documentation, layout and formatting.
- Add code examples in documentation.

To deploy the html version of the documentation run:
`make python-wrapper-gen-docs`
And then use your preferred tool to deploy the create site locally from _site:
(caddy example):
`caddy file-server --listen :2015`

### CODE REVIEW

File: clients/python-wrapper/lakefs/branch.py [hunk 1]
```diff
@@ -32,15 +32,22 @@ def create(self, source_reference_id: str | Reference, exist_ok: bool = False) -
...
         :param source_reference_id: The reference to create the branch from
         :param exist_ok: If False will throw an exception if a branch by this name already exists. Otherwise,
             return the existing branch without creating a new one
         :return: The lakeFS SDK object representing the branch
-        :raises:
-            NotFoundException if repo, branch or source reference id does not exist
-            ConflictException if branch already exists and exist_ok is False
-            NotAuthorizedException if user is not authorized to perform this operation
-            ServerException for any other errors
+        :raise NotFoundException: if repo, branch or source reference id does not exist
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/branch.py:47]: not related to this pr - but I think we should use `repository` and not repo in non-code.

File: clients/python-wrapper/lakefs/branch.py [hunk 2]
```diff
@@ -32,15 +32,22 @@ def create(self, source_reference_id: str | Reference, exist_ok: bool = False) -
         """
         Create a new branch in lakeFS from this object
 
+        Example of creating a new branch:
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/branch.py:35]: Don't we want to place the example with the operation?

File: clients/python-wrapper/lakefs/branch.py [hunk 3]
```diff
@@ -167,11 +169,27 @@ def delete_objects(self, object_paths: str | StoredObject | Iterable[str | Store
         """
         Delete objects from lakeFS
 
+        This method can be used to delete single/multiple objects from branch. It accepts both str and StoredObject
+        types as well as Iterables of these types.
+        Using this method is more performant than sequentially calling delete on objects as it saves the back and forth
+        from the server.
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/branch.py:175]: - Underlaying code uses delete_objects directly and do not chunk the iterables into the server limit. An example where more than 1k object listing will break the delete. Also in terms on memory usage it is not optimal.

We can update the code of delete to handle the limit case and use adapter / lambda to get iterable that will return a path from listing.

Also not part of this PR, we can call delete_objects API on request to delete a single object - the usage pattern that we see in our API also means helps us understand how the system is used.

File: clients/python-wrapper/lakefs/object.py [hunk 4]
```diff
@@ -539,7 +537,22 @@ def path(self) -> str:
...
+            import lakefs
+
+            obj = lakefs.repository("<repository_name>").branch("<branch_name>").object("file.txt")
+            file_size = obj.stat().size_bytes
+
+            with obj.reader(mode='r', pre_sign=True) as fd:
+                # print every other 10 chars
+                while fd.tell() < file_size
+                    print(fd.read(10))
+                    fd.seek(10)
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:555]: seek(10) will not cause to read the same 10 bytes all the time? the default for seek is abs location.

File: clients/python-wrapper/lakefs/object.py [hunk 5]
```diff
@@ -665,13 +679,28 @@ def writer(self,
...
+
+        .. code-block:: python
+
+            import lakefs
+
+            obj = lakefs.repository("<repository_name>").branch("<branch_name>").object("my_image")
+
+            with open("my_local_image", mode='rb') as reader:
+                with obj.writer("wb") as writer:
+                    writer.write(reader.read())
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:692]: I think you can chain with statements:

```suggestion
            with open("my_local_image", mode='rb') as reader, obj.writer("wb") as writer:
                writer.write(reader.read())
```

File: clients/python-wrapper/lakefs/object.py [hunk 6]
```diff
@@ -665,13 +679,28 @@ def writer(self,
...
         :param mode: Write mode - as supported by WriteModes
         :param pre_sign: (Optional), enforce the pre_sign mode on the lakeFS server. If not set, will probe server for
             information.
         :param content_type: (Optional) Specify the data media type
         :param metadata: (Optional) User defined metadata to save on the object
         :return: A Writer object
         """
+        with open("my_file", mode='rb') as reader:
+            with self.writer("wb") as writer:
+                writer.write(reader.read())
```
Comment (nopcoder) [inline clients/python-wrapper/lakefs/object.py:703]: same here (chain with)

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: clients/python-wrapper/lakefs/branch.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:47]: Fixed thanks

File: clients/python-wrapper/lakefs/branch.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:35]: The example is of create branch in the docsting of Branch.create - am I missing something?

File: clients/python-wrapper/lakefs/branch.py (hunk 3 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:175]: Added max_amount to the query. I think the example code should be simple and not over-complex and should not provide a complete solution

File: clients/python-wrapper/lakefs/object.py (hunk 4 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:555]: Fixed, thanks

File: clients/python-wrapper/lakefs/object.py (hunk 6 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/object.py:703]: Added by mistake

File: clients/python-wrapper/lakefs/branch.py (hunk 3 above)
Comment (nopcoder) [reply clients/python-wrapper/lakefs/branch.py:176]: It stand as simple example, but my concern is that usually people learn and use the example for their usage.

Would prefer the library code to handle this case - passing array of 2k items without breaking

**Review by nopcoder — APPROVED** [review]
LGTM with some comment

File: clients/python-wrapper/lakefs/branch.py (hunk 3 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:176]: The limitation is on the server side and might change - I don't think we want to have to maintain this on the SDK side every time we make changes on the server. Let the server handle it and return the error

**Pull request merged**